Refactored the code of the module nmcli

darland opened the pull request:
##### SUMMARY
Refactored the code of the module nmcli.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
nmcli_module

##### ANSIBLE VERSION
```
ansible 2.6.0
```

darland committed:
Refactors nmcli_module

sivel commented:
Can you please explain in detail all of the individual reasons for the refactor?

darland commented:
In the code often encounters a loop that forms a list from the dictionary to add arguments to the command.
```
options = {
    'ip4': self.ip4,
    'gw4': self.gw4,
    'ip6': self.ip6,
    'gw6': self.gw6,
    'autoconnect': self.bool_to_string(self.autoconnect),
    'bridge.ageing-time': self.ageingtime,
    'bridge.forward-delay': self.forwarddelay,
    'bridge.hello-time': self.hellotime,
    'bridge.mac-address': self.mac,
    'bridge.max-age': self.maxage,
    'bridge.priority': self.priority,
    'bridge.stp': self.bool_to_string(self.stp)
}

for key, value in options.items():
     if value is not None:
         cmd.extend([key, value])
```
The loop was moved to a separate method.
```
def _prepare_cmd(self, cmd, options):
    for key in options:
        if self.options[key] is not None:
            cmd.extend([key, self.options[key]])
    return cmd
```

In this case, I replaced the dictionary with a list of arguments, which is passed to this method and iterated through it taking the necessary values from the general dictionary, which is created when initializing instead of variables.


I also rewrote the [method](https://github.com/ansible/ansible/pull/38489/files?diff=split#diff-255f02c063e8277d7ca897265bb36aebL1069) with redundant conditions, simplifying the conditions without losing the logic of the call.

mkrizek reviewed (commented):
I would like to see at least basic set of integration tests to exercise the module, if possible. It's easy to miss something in the changes like this one.

Akasurde commented:
@mkrizek [Tests](https://github.com/ansible/ansible/blob/devel/test/units/modules/net_tools/test_nmcli.py) are already there.

mkrizek commented:
Yeah, I was thinking integration tests though. But if the units covers enough, then I am ok with it.

Akasurde commented:
nmcli modifies network so I am think that will be tricky to cover in the integration tests. btw, changes looks good to me.

darland commented:
It seems to me that need to use `network-integration`, but I don't know how that will run on CI.

Akasurde commented:
@mattclay Could you please advice us on `network-integration` and can we integrate nmcli integration  tests in it ?

mattclay commented:
Modules under `lib/ansible/modules/net_tools/` use `ansible-test integration`.
Modules under `lib/ansible/modules/network/` use `ansible-test network-integration`.

If you want to create integration tests for this module, they belong in: `test/integration/targets/nmcli/`

Akasurde commented:
@darland any news ?

darland committed:
nmcli: Add 'dummy' connection type

darland committed:
nmcli: Add integration tests

darland commented:
Is it possible to run this test without docker?

mattclay commented:
@darland Yes, you can run all of the tests using the various `ansible-test` commands, and use of Docker is optional. If you're not using the `--docker` option I recommend trying `--tox` so you don't need to install all the requirements yourself.

I'm not sure which test(s) you're interested in, but here are the ones relevant to this PR using `tox` and Python 3.6:

```
ansible-test sanity --tox --python 3.6 -v lib/ansible/modules/net_tools/nmcli.py test/integration/targets/nmcli/ test/units/modules/net_tools/test_nmcli.py
ansible-test units --tox --python 3.6 -v nmcli
ansible-test integration --tox --python 3.6 -v --allow-destructive nmcli
```

mattclay commented on test/integration/targets/nmcli/aliases:
```diff
@@ -0,0 +1,5 @@
+posix/ci/group1
+skip/freebsd
+skip/osx
+systemd
```
Why was the `systemd` alias added here? The `systemd` module already a test target.

darland replied:
I thought if I add the systemd to alias the test will run without docker.

mattclay replied:
No, there is no special behavior associated with the `systemd` alias. This line should be removed.

darland commented:
@mattclay thanks. All tests on my local machine passed.
I understand correctly that inside the container, we can not create new network interfaces? If so, how can this test pass to Shippable?

mattclay commented:
@darland I think the easiest thing to do is check for `/.dockerenv` and skipping the tests if found.

Akasurde commented:
@mkrizek Can you please take another look ?

mkrizek reviewed (commented):
Looks good. Though I wonder if we should wait with merging for `2.7` so it's more tested in devel? @Akasurde
